pkg/ipam/allocator.go

```diff
@@ -56,6 +69,13 @@ func (ipam *IPAM) allocateIP(ip net.IP, owner string, pool Pool, needSyncUpstrea
 	ipam.allocatorMutex.Lock()
 	defer ipam.allocatorMutex.Unlock()
 
+	if pool == "" {
```

Comment:
You could add a comment to be explicit that it should not happen.

Reply:
Yeah, so for context for other reviewers: `allocateIP` is usually only called in the context of endpoint restoration, in which point we assume `pool` to be non-empty, so this is basically dead code.

So maybe we should error out here if someone calls `allocateIP` without an explicit pool. Endpoint restoration should call it with an explicit pool (even if it's just `default`). And if the function is called via Cilium API (which I don't think anyone does), there is actually no way to inspect the pool which was selected.

So yeah, discussing this with Donia I'm coming to the conclusion that we probably should just error out if no explicit pool was provided. We can always relax that restriction later.

Reply:
Added an error instead and updated the commit message the reflect that.